BC-4571 - Rewriting the user-group user data deletion method

sszafGCA opened the pull request:
# Description
<!--
-->

## Links to Tickets or other pull requests
- https://ticketsystem.dbildungscloud.de/browse/BC-4571

## Changes
<!--
  What will the PR change?
  Why are the changes required?
  Short notice if a ticket exists, more detailed if not
-->

## Datasecurity
<!--
  Notice about:
  - model changes
  - logging of user data
  - right changes
  - and other user data stuff
  If you are not sure if it is relevant, take a look at confluence or ask the data-security team.
-->

## Deployment
<!--
  Keep in mind to changes to seed data, if changes are done by migration scripts.
  Changes to the infrastructure have to be discussed with the devops

  This point should includes following information:
  - What is required for deployment?
  - Environment variables like FEATURE_XY=true
  - Migration scripts to run, other requirements
-->

## New Repos, NPM pakages or vendor scripts
<!--
  Keep in mind the stability, performance, activity and author.

  Describe why it is needed.
-->

## Approval for review

- [x] DEV: If api was changed - `generate-client:server` was executed in vue frontend and changes were tested and put in a PR with the same branch name.
- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

bn-pass commented on apps/server/src/modules/classes/entity/class-source-options.entity.spec.ts:
```diff
@@ -0,0 +1,31 @@
+import { ClassSourceOptionsEntity } from './class-source-options.entity';
+
+describe('Class Source Options Entity', () => {
```
You can do just _ClassSourceOptionsEntity.name_ so if the class name changes, this string will also be correct without requiring additional refactor.

sszafGCA replied:
fixed

bn-pass commented on apps/server/src/modules/classes/entity/class-source-options.entity.spec.ts:
```diff
@@ -0,0 +1,31 @@
+			return { entity };
+		};
+		it('should throw an error by empty constructor', () => {
+			// @ts-expect-error: Test case
```
Is this really needed? :thinking:

bn-pass commented on apps/server/src/modules/classes/entity/class-source-options.entity.spec.ts:
Please follow the tests naming convention described in [the guidelines](https://hpi-schul-cloud.github.io/schulcloud-server/additional-documentation/nestjs-application/testing.html) .

sszafGCA replied:
fixed

bn-pass commented on apps/server/src/modules/classes/entity/class.entity.spec.ts:
```diff
@@ -0,0 +1,77 @@
+import { ObjectId } from '@mikro-orm/mongodb';
+import { ClassEntity } from './class.entity';
+
+describe('Class Entity', () => {
```
You can do just _ClassEntity.name_ so if the class name changes, this string will also be correct without requiring additional refactor.

bn-pass commented on apps/server/src/shared/testing/factory/class.factory.ts:
Is this entities factory used anywhere else across the project? If not (and in my opinions it shouldn't be used anywhere as we want to isolate the module's entities and expose only the domain objects to the outside), please move it inside the module (e.g. inside the _entity/testing/factory_ submodule).

sszafGCA replied:
fixed

bn-pass commented on apps/server/src/modules/classes/uc/class.uc.ts:
Please delete the UC for now as it should include the authorization layer also and it doesn't have a complete requirements yet - it will be added in another ticket.

sszafGCA replied:
fixed

bn-pass commented on apps/server/src/modules/classes/service/class.service.spec.ts:
```diff
@@ -0,0 +1,91 @@
+				const userId3 = new ObjectId();
+				const class1 = classEntityFactory.withUserIds([userId1, userId2]).build();
+				const class2 = classEntityFactory.withUserIds([userId1, userId3]).build();
+				// class3
```
Some leftover comment?

sszafGCA replied:
fixed

bn-pass commented on apps/server/src/modules/classes/service/class.service.ts:
```diff
@@ -0,0 +1,32 @@
+
+		const updatedEntities: ClassEntity[] = entities.map((entity) => {
+			if (entity.userIds !== undefined) {
+				// eslint-disable-next-line @typescript-eslint/no-shadow
```
Please try to remove this annotation.

sszafGCA replied:
fixed

wolfganggreschus commented on apps/server/src/modules/classes/class.module.ts:
I'm a little bit confused that you named the module classes instead of class. A new convention?

sszafGCA replied:
Module is called class look at line 9

Metauriel replied:
he is right about the module folder though, all other folders are named singular.

sszafGCA replied:
ohh okay fixed it

sszafGCA commented:
> For the future, we want to keep persistance-specific code (entities) out of the business logic as much as possible. Because of this, we have introduced a seperation of entities and dos in recent modules.
> 
> The rule we agreed on, is that the repos should always return and accept dos, and hide the entities within their details. how the mapping happens between dos and entities is left as implementation detail to the repo.
> 
> In this case it might seem like a slight overengineering, but since im not sure what developments will happen in the future on the class module, and more importantly to speed up migrating the rest of class related code from feathers to nest, I would prefer you to already create a ClassDo and make sure the repository returns it.
> 
> Beyond that, your code looks really good!

Hi Thomas I have added domain layer. Could you take a look again?

Metauriel commented on apps/server/src/modules/classes/service/class.service.ts:
```diff
@@ -0,0 +1,28 @@
+
+		const updatedClasses: Class[] = domainObjects.map((domainObject) => {
+			if (domainObject.userIds !== undefined) {
+				domainObject.userIds.filter((userId1) => userId1 !== userId);
```
since userIds only has a getter, I would not expect this to work. 
It does in this case, but userIds mit as well return a copy to protect itself from changes.

There are a couple of ways to change this. My personal favorite are functions to set data (either updateUsers(), or even removeUsers(), but a setter is possible as well.

sszafGCA replied:
fixed

Metauriel commented on apps/server/src/modules/classes/service/mapper/class.mapper.ts:
```diff
@@ -0,0 +1,50 @@
+import { ClassSourceOptions } from '../../domain/class-source-options.do';
+import { ClassEntity } from '../../entity';
+
+export class ClassMapper {
```
this should be moved into the repo folder, as its used (only) there.

sszafGCA replied:
fixed
